packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart

```diff
@@ -0,0 +1,342 @@
+/// Defaults to using SSH port forwarding.
+PortForwardingFunction fuchsiaPortForwardingFunction = _SshPortForwarder.start;
+
+/// Sets [fuchsiaPortForwardingFunction] back to the default SSH port forwarding
+/// implementation.
+void restoreFuchsiaPortForwardingFunction() {
+  fuchsiaPortForwardingFunction = _SshPortForwarder.start;
+}
+
+/// Manages a remote connection to a Fuchsia Device.
+///
+/// Provides affordances to observe and connect to Flutter views, isolates, and
+/// perform actions on the Fuchsia device's various VM services.
+///
+/// Note that this class can be connected to several instances of the Fuchsia
+/// device's Dart VM at any given time.
+class FuchsiaRemoteConnection {
+  final List<PortForwarder> _forwardedVmServicePorts = <PortForwarder>[];
+  final SshCommandRunner _sshCommandRunner;
+  final bool _useIpV6Loopback;
+
+  /// VM service cache to avoid repeating handshakes across function
+  /// calls. Keys a forwarded port to a DartVm connection instance.
+  final Map<int, DartVm> _dartVmCache = <int, DartVm>{};
+
+  FuchsiaRemoteConnection._(this._useIpV6Loopback, this._sshCommandRunner);
+
+  /// Same as [FuchsiaRemoteConnection.connect] albeit with a provided
+  /// [SshCommandRunner] instance.
+  @visibleForTesting
+  static Future<FuchsiaRemoteConnection> connectWithSshCommandRunner(
+      SshCommandRunner commandRunner) async {
+    final FuchsiaRemoteConnection connection = new FuchsiaRemoteConnection._(
+        isIpV6Address(commandRunner.address), commandRunner);
+    await connection._forwardLocalPortsToDeviceServicePorts();
+    return connection;
+  }
+
+  /// Opens a connection to a Fuchsia device.
+  ///
+  /// Accepts an `address` to a Fuchsia device, and optionally a `sshConfigPath`
+  /// in order to open the associated ssh_config for port forwarding.
+  ///
+  /// Will throw an [ArgumentError] if `address` is malformed.
+  ///
+  /// Once this function is called, the instance of [FuchsiaRemoteConnection]
+  /// returned will keep all associated DartVM connections opened over the
+  /// lifetime of the object.
+  ///
+  /// At its current state Dart VM connections will not be added or removed over
+  /// the lifetime of this object.
+  ///
+  /// Throws an [ArgumentError] if the supplied `address` is not valid IPv6 or
+  /// IPv4.
+  ///
+  /// Note that if `address` is ipv6 link local (usually starts with fe80::),
+  /// then `interface` will probably need to be set in order to connect
+  /// successfully (that being the outgoing interface of your machine, not the
+  /// interface on the target machine).
+  static Future<FuchsiaRemoteConnection> connect(String address,
+      [String interface = '', String sshConfigPath]) async {
+    return await FuchsiaRemoteConnection.connectWithSshCommandRunner(
+        new SshCommandRunner(
+            address: address,
+            interface: interface,
+            sshConfigPath: sshConfigPath));
```

Comment:
if you use dartfmt, add trailing commas before each `)` and `]` and rewrap this.

It should look like this to match Flutter style:

```dart
static Future<FuchsiaRemoteConnection> connect(
  String address, [
  String interface = '',
  String sshConfigPath,
]) async {
  return await FuchsiaRemoteConnection.connectWithSshCommandRunner(
    new SshCommandRunner(
      address: address,
      interface: interface,
      sshConfigPath: sshConfigPath,
    ),
  );
}
```

Reply:
Fixed.